### PR TITLE
Add foreign keys to sequencing metrics model

### PR DESCRIPTION
**Pull request by seallard**
## Description
This PR updates the SampleLaneSequencingMetrics table with foreign keys to the sample and flow cell tables.

Specifically linking flow_cell_name to Flowcell.name and sample_internal_id to Sample.internal_id. This enforces the relationship that each SampleLaneSequencingMetrics record corresponds to a particular Flowcell and Sample.

New relationships have also been added to the Flowcell and Sample models, allowing easier access from these models to their associated SampleLaneSequencingMetrics.

### Added
- Foreign keys on SampleLaneSequencingMetrics

### This [version](https://semver.org/) is a
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: alembic/versions/9d1483e638af_add_foreign_keys_to_.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+down_revision = "f5e0db62a5a7"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # Add foreign key constraints to SampleLaneSequencingMetrics table
+    op.add_column(
+        "sample_lane_sequencing_metrics",
+        sa.Column("flow_cell_name", sa.String(length=128), nullable=False),
```
Comment (ChrOertlin) [inline alembic/versions/9d1483e638af_add_foreign_keys_to_.py:23]: This column already exists, not sure if we just need to change it to a foreign key

File: alembic/versions/9d1483e638af_add_foreign_keys_to_.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    )
+    op.create_foreign_key(
+        None, "sample_lane_sequencing_metrics", "flowcell", ["flow_cell_name"], ["name"]
+    )
+    op.add_column(
+        "sample_lane_sequencing_metrics",
+        sa.Column("sample_internal_id", sa.String(length=128), nullable=False),
```
Comment (ChrOertlin) [inline alembic/versions/9d1483e638af_add_foreign_keys_to_.py:30]: same here

File: cg/store/models.py [hunk 3]
```diff
@@ -519,6 +519,9 @@ class Flowcell(Model):
     updated_at = Column(types.DateTime, onupdate=dt.datetime.now)
 
     samples = orm.relationship("Sample", secondary=flowcell_sample, backref="flowcells")
+    sample_lane_sequencing_metrics = orm.relationship(
+        "SampleLaneSequencingMetrics", backref="flowcell", lazy="dynamic"
```
Comment (ChrOertlin) [inline cg/store/models.py:523]:
```suggestion
        "SampleLaneSequencingMetrics", back_populates="flowcell", lazy="dynamic"
```

backref is a legacy parameter:
https://docs.sqlalchemy.org/en/20/orm/backref.html

**Review by ChrOertlin — APPROVED** [review]
some minor things we should consider

**Review by henrikstranneheim — APPROVED** [review]
Great work 👍

File: alembic/versions/9d1483e638af_add_foreign_keys_to_.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+Revises: f5e0db62a5a7
+Create Date: 2023-06-01 13:40:25.835471
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+# revision identifiers, used by Alembic.
+revision = "9d1483e638af"
+down_revision = "f5e0db62a5a7"
```
Comment (ChrOertlin) [inline alembic/versions/9d1483e638af_add_foreign_keys_to_.py:13]:
```suggestion
down_revision = "e6a3f1ad4b50"
```

File: alembic/versions/9d1483e638af_add_foreign_keys_to_.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+"""Add foreign keys to SampleLaneSequencingMetrics
+
+Revision ID: 9d1483e638af
+Revises: f5e0db62a5a7
```
Comment (ChrOertlin) [inline alembic/versions/9d1483e638af_add_foreign_keys_to_.py:4]:
```suggestion
Revises: e6a3f1ad4b50
```

File: cg/store/models.py [hunk 6]
```diff
@@ -766,10 +769,12 @@ class SampleLaneSequencingMetrics(Model):
     __tablename__ = "sample_lane_sequencing_metrics"
 
     id = Column(types.Integer, primary_key=True)
-    flow_cell_name = Column(types.String(128), nullable=False)
+    flow_cell_name = Column(types.String(128), ForeignKey("flowcell.name"), nullable=False)
```
Comment (ChrOertlin) [inline cg/store/models.py:772]:
```suggestion
    flow_cell_name = Column(types.String(32), ForeignKey("flowcell.name"), nullable=False)
```

File: cg/store/models.py [hunk 7]
```diff
@@ -766,10 +769,12 @@ class SampleLaneSequencingMetrics(Model):
     __tablename__ = "sample_lane_sequencing_metrics"
 
     id = Column(types.Integer, primary_key=True)
-    flow_cell_name = Column(types.String(128), nullable=False)
+    flow_cell_name = Column(types.String(32), ForeignKey("flowcell.name"), nullable=False)
+    flow_cell = orm.relationship(Flowcell, back_populates="sequencing_metrics")
     flow_cell_lane_number = Column(types.Integer)
 
-    sample_internal_id = Column(types.String(128), nullable=False)
+    sample_internal_id = Column(types.String(128), ForeignKey("sample.internal_id"), nullable=False)
```
Comment (ChrOertlin) [inline cg/store/models.py:776]:
```suggestion
    sample_internal_id = Column(types.String(32), ForeignKey("sample.internal_id"), nullable=False)
```

**Pull request closed without merging**